src/Toml/PrefixTree.hs

```diff
@@ -0,0 +1,109 @@
+       , lookupT
+       , deleteT
+
+         -- * Types
+       , Piece (..)
+       , Key (..)
+       , Prefix
+       , KeysDiff (..)
+       ) where
+
+import Prelude hiding (lookup)
+
+import Data.Hashable (Hashable)
+import Data.HashMap.Strict (HashMap)
+import Data.List.NonEmpty (NonEmpty)
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.Text (Text)
+
+import qualified Data.HashMap.Strict as HashMap
+
+-- | Represents the key piece of some layer.
+newtype Piece = Piece { unPiece :: Text }
+    deriving (Show, Eq, Hashable)
+
+-- | Full representation of Key contains pieces of every layer.
+newtype Key = Key { unKey :: NonEmpty Piece }
+    deriving (Show, Eq)
+
+-- | Type synonym for 'Key'.
+type Prefix = Key
+
+-- | Map of layer names and corresponding 'PrefixTree's.
+type PrefixMap a = HashMap Piece (PrefixTree a)
+
+-- | Data structure to represent table tree for @toml@.
+data PrefixTree a
+    = Leaf !Key !a
+    | Branch { bCommonPref :: !Prefix         -- ^ greatest common prefix
+             , bVal        :: !(Maybe a)      -- ^ value by key = prefix
+             , bPrefixMap  :: !(PrefixMap a)  -- ^ suffixes of prefix
+             }
+
+data KeysDiff
+      -- | Keys are equal
+    = Equal
+      -- | Keys don't have any common part.
+    | NoPrefix
+      -- | The first key is the prefix for the second one.
+    | FstIsPref { diff :: !Key}
+      -- | The second key is the prefix for the first one.
+    | SndIsPref { diff :: !Key}
+      -- | Key have same prefix.
+    | Diff { pref    :: !Key
+           , diffFst :: !Key
+           , diffSnd :: !Key
+           }
+    deriving (Show, Eq)
+
+toKey :: Piece -> [Piece] -> Key
+toKey x xs = Key $ x :| xs
+
+keysDiff :: Key -> Key -> KeysDiff
+keysDiff (Key (x :| xs)) (Key (y :| ys))
+    | x == y    = listSame xs ys []
+    | otherwise = NoPrefix
+  where
+    listSame :: [Piece] -> [Piece] -> [Piece] -> KeysDiff
+    listSame [] []     _ = Equal
+    listSame [] (s:ss) _ = FstIsPref $ toKey s ss
+    listSame (f:fs) [] _ = SndIsPref $ toKey f fs
+    listSame (f:fs) (s:ss) pr =
+        if f == s
+        then listSame fs ss (pr ++ [f])
+        else Diff (toKey x pr) (toKey f fs) (toKey s ss)
+
+-- | Creates a 'PrefixTree' of one key-value element.
+singleT :: Key -> a -> PrefixTree a
+singleT = Leaf
+
+-- | Inserts key-value element into the given 'PrefixTree'.
+insertT :: Key -> a -> PrefixTree a -> PrefixTree a
+insertT = undefined
+
+-- | Looks up the value at a key in the 'PrefixTree'.
+lookupT :: Key -> PrefixTree a -> Maybe a
+lookupT lk (Leaf k v) = if lk == k then Just v else Nothing
+lookupT lk (Branch pref mv prefMap) =
+    case keysDiff pref lk of
+        Equal -> mv
+        NoPrefix -> Nothing
+        Diff _ _ _ -> Nothing
+        SndIsPref _ -> Nothing
+        FstIsPref (Key (r :| rs)) -> HashMap.lookup r prefMap >>= \pt ->
+            case rs of
+                []     -> getVal pt
+                (x:xs) -> lookupT (toKey x xs) pt
```

Comment:
I think this case is implemented slightly incorrect. It's hard to track all these prefixes properly... But I think the main problem here is that you're trying to optimize things and for key `x :| xs` store `x` in `HashMap` head and then `xs` in the rest of the tree. Unfortunately, this won't work. Consider this mapping:

```toml
a.x = 1
a.y = 2
```

The only way to store this in our `PrefixTree` is the following:

```haskell
Branch
  { bCommonPrefix = "a" :| []
  , bVal = Nothing
  , bPrefixMap =
       [ "x" → Leaf ("x" :| []) 1
       , "b" → Leaf ("b" :| []) 2
       ]
  }
```

So yes, we will have some data duplication. We will store head of each key in `HashMap` twice: as a key in `HashMap` and as a first component of `Prefix` in `PrefixTree` value. This `HashMap` is needed only for performance reasons (that's why we don't store list of `PrefixTree`s). And this cost is not really that big. But it simplifies implementation a lot.